### PR TITLE
[PLGN-299] theHive error handling

### PR DESCRIPTION
**Pull request by cmcnally-r7**
### Description

Describe the proposed changes:

  - Add further error handling to produce useful messages when the response falls outside of the defined status codes.
  - Changed `create_case` to not include a task input within the case. 
  - focus on `api.py` and `create_case` action.py

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/thehive/komand_thehive/util/api.py [hunk 1]
```diff
@@ -89,9 +89,10 @@ def _call_api(
             raise PluginException(preset=PluginException.Preset.UNAUTHORIZED, data=response)
         if response.status_code == 404:
             raise PluginException(preset=PluginException.Preset.NOT_FOUND, data=response)
-        if response.status_code == 500:
+        if response.status_code >= 500:
             raise PluginException(preset=PluginException.Preset.SERVER_ERROR, data=response)
         if response.status_code in (200, 201):
             return response.json()
         if response.status_code == 204:
```
Comment (ablakley-r7) [inline plugins/thehive/komand_thehive/util/api.py:96]: I would recommend accounting for all 200-299 responses positively as they are all successful and shouldn't raise an error.

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]

File: plugins/thehive/komand_thehive/util/api.py (hunk 1 above)
Comment (cmcnally-r7) [reply plugins/thehive/komand_thehive/util/api.py:96]: The previous code we had:
```
if 200 >= response.status_code < 300:
```
caused a lot of issues. 

I'll retry it 👍🏻

Comment (cmcnally-r7) [reply plugins/thehive/komand_thehive/util/api.py:96]: Yeah so this was the error I was getting when trying to cover whole range with the new Exception that is raised at the end
```
Something unexpected occurred. Check the logs and if the issue persists please contact support. Response was: <Response [201]>
{"body": {"log": "", "status": "error", "meta": {}, "error": "An error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: <Response [201]>", "exception": {"cause": "Something unexpected occurred.", "assistance": "Check the logs and if the issue persists please contact support.", "data": "<Response [201]>"}}, "version": "v1", "type": "action_event"}
```

I'm probably just not noticing something really silly here. Log above to keep track ^

Comment (ablakley-r7) [reply plugins/thehive/komand_thehive/util/api.py:96]: I would try if response.status_code in range(200, 300)

Comment (joneill-r7) [reply plugins/thehive/komand_thehive/util/api.py:94]: Would it be better to change the ordering so the happy path is checked first to save checking the error codes when it's (hopefully) unlikely to hit these? 

Also is it better to do `data=response.content` in the exception? if we return `Response was <Response 400>` etc it doesn't tell us much but response.content should include the error from the endpoint, the cause/assistance should let us know which status code. might be an issue for `Preset.UNKNOWN` as the exception won't then include the status code, unless an additional logger can be added for that?

Comment (cmcnally-r7) [reply plugins/thehive/komand_thehive/util/api.py:94]: The order does not matter so much but I've switch them around anyway.
Good catch with the `response.content`. Changing the values now

**Review by ablakley-r7 — APPROVED** [review]

**Review by joneill-r7 — APPROVED** [review]
lgtm

File: plugins/thehive/bin/komand_thehive [hunk 2]
```diff
@@ -34,27 +34,27 @@ def main():
...
+        
 
     """Run plugin"""
     cli = insightconnect_plugin_runtime.CLI(ICONThehive())
     cli.run()
 
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (TKuzas-R7) [inline plugins/thehive/bin/komand_thehive:60]:
```suggestion
    main()
    
```

File: plugins/thehive/komand_thehive/actions/close_case/schema.py [hunk 3]
```diff
@@ -61,9 +62,10 @@ class CloseCaseOutput(insightconnect_plugin_runtime.Output):
...
     "success"
-  ]
+  ],
+  "definitions": {}
 }
     """)
 
     def __init__(self):
-        super(self.__class__, self).__init__(self.schema)
+        super(self.__class__, self).__init__(self.schema)
```
Comment (TKuzas-R7) [inline plugins/thehive/komand_thehive/actions/close_case/schema.py:71]:
```suggestion
        super(self.__class__, self).__init__(self.schema)
        
```

**Review by TKuzas-R7 — APPROVED** [review]
Looks good but theres a lot of trailing new lines missing.
Don't know if thats important or not though.

**Comment by cmcnally-r7** [comment]
> Looks good but theres a lot of trailing new lines missing. Don't know if thats important or not though.

I think we can safely ignore for now and make a note to fix this issue in the future.

**Pull request merged**